### ISSUE TITLE
identity_op warns for bitflag constant

### ISSUE THREAD

**Issue by konstin** [rust-lang/rust-clippy#3430]
[pyo3](https://github.com/PyO3/pyo3) has to map certain constants and flags from the python headers, which includes the code below:

```rust
// Flag bits for printing:
pub const Py_PRINT_RAW: c_int = 1; // No string quotes etc.

// PyBufferProcs contains bf_getcharbuffer
pub const Py_TPFLAGS_HAVE_GETCHARBUFFER: c_long = (1 << 0);

// PySequenceMethods contains sq_contains
pub const Py_TPFLAGS_HAVE_SEQUENCE_IN: c_long = (1 << 1);

// PySequenceMethods and PyNumberMethods contain in-place operators
pub const Py_TPFLAGS_HAVE_INPLACEOPS: c_long = (1 << 3);

// PyNumberMethods do their own coercion
pub const Py_TPFLAGS_CHECKTYPES: c_long = (1 << 4);
```

This code raises a warning:

```
warning: the operation is ineffective. Consider reducing it to `1`
   --> src/ffi2/object.rs:676:51
    |
676 | pub const Py_TPFLAGS_HAVE_GETCHARBUFFER: c_long = (1 << 0);
    |                                                   ^^^^^^^^
    |
    = note: #[warn(clippy::identity_op)] on by default
```

The explanation says:

> This code can be removed without changing the meaning. So it just obscures what’s going on. Delete it mercilessly.

IMHO `(1 << 0)` is not obscuring the code, but relevant for pointing out that this constant is the first entry in a bitfield. Therefore this is imho a false positive in clippy.

`cargo clippy -V`: clippy 0.0.212 (a20599a 2018-11-01)

**Comment by konstin** [comment]
It's for sure an identity op, but I'd argue that it's not bad style or erroneous, but an important note for what that constant does. It would also be inconsistent to use `x << 0` everywhere else but not on the first element.

**Comment by flip1995** [comment]
I would suggest to allow `(_ << 0)` in constant definitions. But I think generally not linting it would be a bad idea.

**Comment by konstin** [comment]
I agree

**Comment by oli-obk** [comment]
so... generally `constant << 0` or just in constant contexts?

**Comment by konstin** [comment]
Good question. All seems like a good choice because `extern_c_function((1 << 0) | (1 << 2))` would be reasonble, otoh you could argue you should define constants for that.

**Comment by khoek** [comment]
I've just run into this in several places writing code for an embedded target.

I agree, I even think the pattern `1 << 0` verbatim is so common in bit manipulations (compared even to all variants of `MASK << 0`) that it would be worth it to allow.
